flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -650,14 +658,48 @@ private TaskManagerSlot createAndRegisterTaskManagerSlot(SlotID slotId, Resource
 	@Nullable
 	private PendingTaskManagerSlot findExactlyMatchingPendingTaskManagerSlot(ResourceProfile resourceProfile) {
 		for (PendingTaskManagerSlot pendingTaskManagerSlot : pendingSlots.values()) {
-			if (pendingTaskManagerSlot.getResourceProfile().equals(resourceProfile)) {
+			if (isPendingSlotExactlyMatchingResourceProfile(pendingTaskManagerSlot, resourceProfile)) {
 				return pendingTaskManagerSlot;
 			}
 		}
 
 		return null;
 	}
 
+	private boolean isPendingSlotExactlyMatchingResourceProfile(PendingTaskManagerSlot pendingTaskManagerSlot, ResourceProfile resourceProfile) {
+		return pendingTaskManagerSlot.getResourceProfile().equals(resourceProfile);
+	}
+
+	private boolean isMaxSlotNumExceededAfterRegistration(SlotReport initialSlotReport) {
+		final int numReportedNewSlots = initialSlotReport.getNumSlotStatus();
+		final int numRegisteredSlots =  getNumberRegisteredSlots();
+		final int numPendingSlots = getNumberPendingTaskManagerSlots();
+
+		// check if the total number exceed before matching pending slot.
+		if (numRegisteredSlots + numPendingSlots + numReportedNewSlots <= maxSlotNum) {
+			return false;
+		}
+
+		// check how many exceed slots could be consumed by pending slot.
+		final int totalSlotNum = numRegisteredSlots + numPendingSlots + getNumNonPendingReportedNewSlots(initialSlotReport);
+		return totalSlotNum > maxSlotNum;
+	}
+
+	private int getNumNonPendingReportedNewSlots(SlotReport slotReport) {
+		final Set<TaskManagerSlotId> matchingPendingSlots = new HashSet<>();
+
+		for (SlotStatus slotStatus : slotReport) {
+			for (PendingTaskManagerSlot pendingTaskManagerSlot : pendingSlots.values()) {
+				if (!matchingPendingSlots.contains(pendingTaskManagerSlot.getTaskManagerSlotId()) &&
```

Comment:
why do we need `!matchingPendingSlots.contains(pendingTaskManagerSlot.getTaskManagerSlotId())`? it does not look like `SlotReport` can contain duplicated `TaskManagerSlotId`.

Reply:
This function checks how many `pendingTaskManagerSlot` could be fulfilled by the slot in `SlotReport`. If we do not record which `pendingTaskManagerSlot` has already been matched before, all the slot in `SlotReport` would match with a single `pendingTaskManagerSlot`, which is not correct.

Reply:
true, sorry, I confused it with `slotReport`